lang/query/src/cache.rs

```diff
@@ -0,0 +1,75 @@
+#[cfg(not(target_arch = "wasm32"))]
+use std::time;
+#[cfg(target_arch = "wasm32")]
+use web_time as time;
+
+use url::Url;
+
+use ast::HashMap;
+
+use crate::Database;
+
+type IsStaleFn = dyn Fn(&Database, &Url, time::Instant) -> bool + Send + Sync;
```

Comment:
I am not really happy with the mutual recursion between `Cache` and `Database`, because this makes it super hard to reason about how cache invalidation works. Can we have an alternative design where `stale` is a boolean in `CacheEntry`, and we have an API function `mark_as_stale(&mut self, url: Url)` in the cache which is called by the Database? In general I would prefer not to use time-base cache invalidation, but something along the dirty-bit approach described in https://www.microsoft.com/en-us/research/uploads/prod/2018/03/build-systems.pdf for example.